src/mlir/dialect/VeronaTypes.cc

```diff
@@ -0,0 +1,334 @@
+    return Base::get(ctx, VeronaTypes::Meet, elements);
+  }
+
+  llvm::ArrayRef<mlir::Type> MeetType::getElements() const
+  {
+    return getImpl()->elements;
+  }
+
+  JoinType JoinType::get(MLIRContext* ctx, llvm::ArrayRef<mlir::Type> elements)
+  {
+    assert(areVeronaTypes(elements));
+    return Base::get(ctx, VeronaTypes::Join, elements);
+  }
+
+  llvm::ArrayRef<mlir::Type> JoinType::getElements() const
+  {
+    return getImpl()->elements;
+  }
+
+  IntegerType IntegerType::get(MLIRContext* ctx, size_t width, unsigned sign)
+  {
+    return Base::get(ctx, VeronaTypes::Integer, width, sign);
+  }
+
+  size_t IntegerType::getWidth() const
+  {
+    return getImpl()->width;
+  }
+
+  bool IntegerType::getSign() const
+  {
+    return getImpl()->sign;
+  }
+
+  CapabilityType CapabilityType::get(MLIRContext* ctx, Capability cap)
+  {
+    return Base::get(ctx, VeronaTypes::Capability, cap);
+  }
+
+  Capability CapabilityType::getCapability() const
+  {
+    return getImpl()->capability;
+  }
+
+  /// Parse a list of types, surrounded by angle brackets and separated by
+  /// commas. The types inside the list must be Verona types and should not use
+  /// the `!verona.` prefix.
+  ///
+  /// Empty lists are allowed, but must still use angle brackets, i.e. `< >`.
+  /// Lists of one elements are also allowed.
+  static ParseResult
+  parseTypeList(DialectAsmParser& parser, llvm::SmallVectorImpl<Type>& result)
+  {
+    if (parser.parseLess())
+      return failure();
+
+    if (succeeded(parser.parseOptionalGreater()))
+      return success();
+
+    do
+    {
+      mlir::Type element = parseVeronaType(parser);
+      if (!element)
+        return failure();
+
+      result.push_back(element);
+    } while (succeeded(parser.parseOptionalComma()));
+
+    if (parser.parseGreater())
+      return failure();
+
+    return success();
+  }
+
+  static Type parseMeetType(MLIRContext* ctx, DialectAsmParser& parser)
+  {
+    SmallVector<mlir::Type, 2> elements;
+    if (parseTypeList(parser, elements))
+      return Type();
+    return MeetType::get(ctx, elements);
+  }
+
+  static Type parseJoinType(MLIRContext* ctx, DialectAsmParser& parser)
+  {
+    SmallVector<mlir::Type, 2> elements;
+    if (parseTypeList(parser, elements))
+      return Type();
+    return JoinType::get(ctx, elements);
+  }
+
+  static Type parseIntegerType(
+    MLIRContext* ctx, DialectAsmParser& parser, StringRef keyword)
+  {
+    size_t width = 0;
+    if (keyword.substr(1).getAsInteger(10, width))
+    {
+      parser.emitError(parser.getNameLoc(), "unknown verona type: ") << keyword;
+      return Type();
+    }
+    bool sign = keyword.startswith("S");
+    return IntegerType::get(ctx, width, sign);
+  }
+
+  Type parseVeronaType(DialectAsmParser& parser)
+  {
+    MLIRContext* ctx = parser.getBuilder().getContext();
+
+    StringRef keyword;
+    if (parser.parseKeyword(&keyword))
+      return Type();
+
+    if (keyword == "meet")
```

Comment:
In the parser and generator, we're using the `str2tag` to make that into a trivial integer switch.

Reply:
Interesting, thanks. I don't really want to tie the `dialect/` folder to the peg library, and duplicating the implementation feels overkill for just this switch statement.

There's also `StringSwitch` from LLVM (which I haven't used before), but it seems to require values to be associated to each case, not blocks of code. Obviously function pointers and lambdas are values, so it could work, but it feels a bit too complicated.

Reply:
> Interesting, thanks. I don't really want to tie the `dialect/` folder to the peg library, and duplicating the implementation feels overkill for just this switch statement.

Good point.

> There's also `StringSwitch` from LLVM (which I haven't used before), but it seems to require values to be associated to each case, not blocks of code. Obviously function pointers and lambdas are values, so it could work, but it feels a bit too complicated.

You could outline a function that turns strings into enum values and use the switch on those values.

Reply:
Are you doing the `StringSwitch` on this PR or leaving for later?

Reply:
anyway is fine. This PR looks good to me now, thanks!